modules/internal/ChapelArray.chpl

```diff
@@ -2939,13 +2991,189 @@ module ChapelArray {
   }
 
   @chpldoc.nodoc
+  @edition(last="2.0")
   proc reshape(A: _iteratorRecord, D: domain) {
     if !D.isRectangular() then
-      compilerError("reshape(A,D) is meaningful only when D is a rectangular domain; got D: ", D.type:string);
+      compilerError("reshape() only currently supports rectangular domains; got D: ", D.type:string);
     var B = for (i,a) in zip(D, A) do a;
     return B;
   }
 
+  // The following is the proposed new reshape() implementation that
+  // supports aliasing by default, currently part of a pre-edition
+
+  @chpldoc.nodoc
+  config param checkReshapeDimsByDefault = boundsChecking;
+
+  // The following overloads take a varargs list of ranges.  The fact
+  // that there are three distinct overloads is regrettable, and a
+  // result of working around:
+  //   https://github.com/chapel-lang/chapel/issues/17188
+
+  pragma "no promotion when by ref"
+  pragma "fn returns aliasing array"
+  @chpldoc.nodoc
+  @edition(first="pre-edition")
+  proc reshape(arr: [], ranges: range(?)...) {
+    if ranges.size == 1 && ranges(0).bounds == boundKind.low {
```

Comment:
Even though the signatures must differ, is it possible to have these call out to a more verbose implementation procedure that takes all the flags/configuration arguments _before_ the varargs? That way we could perhaps reduce redundancy at least a little bit.

Reply:
#27312 does some additional cleanup and refactoring to reduce some redundancy.  However, I'm not sure that I can do more here without a fix to https://github.com/chapel-lang/chapel/issues/17188.  Let me know if you see something that I'm missing.

Reply:
Looks good, maybe we can address 17188 in dyno.